File: matlab/src/+pme/weights.m
function [W, stats] = weights(delta, layout, cfg, Uinfo, blocks)
%PME_WEIGHTS Build block-diagonal weights consistent with "ninf as #information".
%
% Idea:
% - Each INFORMATION contributes ~1 to trace(W*Cov) => sum(eigs) ~ ninf.
% - Geometry D counts as 1 info (when present in P).
% - Each FIELD condition counts as 1 info (weight per condition block).
% - Each SCALAR (and each scalar condition) counts as 1 info (weight per row).
% - Variables U weight is usually 0 in PI/PD/PME (as in your scripts).

mode = lower(string(cfg.mode));

Dlen = layout.D.nRows;
Ulen = Uinfo.Mact;
Flen = size(blocks.F,1);
Clen = size(blocks.C,1);

% Helper: safe inverse
invv = @(x) 1/max(x, eps);

switch mode
    case "pme"
        % P=[D;U]
        N = Dlen + Ulen;
        W = zeros(N,N);

        % D = 1 information
        varD = sum(var(delta(1:Dlen,:),1,2));
        wD = invv(varD);
        W(1:Dlen,1:Dlen) = wD * eye(Dlen);

        % U usually not weighted in PME PCA (keep 0)
        % If you ever want it: set wU = invv(sum(var(...)))
        wU = 0;
        if wU > 0
            W(Dlen+1:Dlen+Ulen, Dlen+1:Dlen+Ulen) = wU * eye(Ulen);
        end

        stats = pack_stats(mode, Dlen, Ulen, Flen, Clen, 1, 0, 0);

    case "pi"
        % P=[D;U;F;C]
        N = Dlen + Ulen + Flen + Clen;
        W = zeros(N,N);

        % --- D = 1 information ---
        ptr = 1;
        varD = sum(var(delta(ptr:ptr+Dlen-1,:),1,2)); ptr = ptr + Dlen;
        wD = invv(varD);
        W(1:Dlen,1:Dlen) = wD * eye(Dlen);

        % --- U typically 0 ---
        % still advance ptr across U block in delta indexing
        varU = sum(var(delta(ptr:ptr+Ulen-1,:),1,2)); %#ok<NASGU>
        ptrU0 = ptr;
        ptr = ptr + Ulen;
        wU = 0;
        if wU > 0
            W(Dlen+1:Dlen+Ulen, Dlen+1:Dlen+Ulen) = wU * eye(Ulen);
        end

        % --- F: 1 info per field condition ---
        [WF, nFinfo, ptr] = build_WF(delta, ptr, Dlen+Ulen, layout, Flen);
        if ~isempty(WF)
            a = Dlen+Ulen+1; b = Dlen+Ulen+Flen;
            W(a:b, a:b) = WF;
        end

        % --- C: 1 info per scalar condition (row-by-row) ---
        [WC, nCinfo] = build_WC(delta, ptr, Dlen+Ulen+Flen, layout, Clen);
        if ~isempty(WC)
            a = Dlen+Ulen+Flen+1; b = Dlen+Ulen+Flen+Clen;
            W(a:b, a:b) = WC;
        end

        stats = pack_stats(mode, Dlen, Ulen, Flen, Clen, 1, nFinfo, nCinfo);
        stats.ptrU0 = ptrU0; % debug

    case "pd"
        % P=[U;F;C]
        N = Ulen + Flen + Clen;
        W = zeros(N,N);

        % --- U weight 0 (physics-driven) ---
        wU = 0;
        if wU > 0
            W(1:Ulen,1:Ulen) = wU * eye(Ulen);
        end

        ptr = 1 + Ulen;

        % --- F: 1 info per field condition ---
        [WF, nFinfo, ptr] = build_WF(delta, ptr, Ulen, layout, Flen);
        if ~isempty(WF)
            a = Ulen+1; b = Ulen+Flen;
            W(a:b, a:b) = WF;
        end

        % --- C: 1 info per scalar condition ---
        [WC, nCinfo] = build_WC(delta, ptr, Ulen+Flen, layout, Clen);
        if ~isempty(WC)
            a = Ulen+Flen+1; b = Ulen+Flen+Clen;
            W(a:b, a:b) = WC;
        end

        if Flen==0
            warning('PD-PME with no F: physics-driven DR limited by C.');
        end

        stats = pack_stats(mode, Dlen, Ulen, Flen, Clen, 0, nFinfo, nCinfo);

    otherwise
        error('Unknown mode: %s', mode);
end
end

% ================= helpers =================

function [WF, nFinfo, ptr_out] = build_WF(delta, ptr_in, offset, layout, Flen)
% Build weights for F block: 1 info per field condition (block per condition)
% Returns:
%  WF: [Flen x Flen] diagonal (constant per condition block)
%  nFinfo: number of field informations
%  ptr_out: updated ptr in delta indexing

WF = [];
nFinfo = 0;
ptr = ptr_in;

if Flen <= 0
    ptr_out = ptr;
    return;
end

WF = zeros(Flen, Flen);

if isfield(layout,'F') && isfield(layout.F,'items') && ~isempty(layout.F.items)
    local = 1;
    for i=1:numel(layout.F.items)
        it = layout.F.items(i);
        nCond = it.nCond;
        nRows = it.nRows;
        K = nRows / max(nCond,1);

        for c=1:nCond
            rr_delta = ptr:(ptr+K-1);
            varBlock = sum(var(delta(rr_delta,:),1,2));
            w = 1/max(varBlock, eps);

            rr_W = local:(local+K-1);
            WF(rr_W, rr_W) = w * eye(K);

            ptr = ptr + K;
            local = local + K;
            nFinfo = nFinfo + 1;
        end
    end
else
    % fallback: treat whole F as 1 info
    varF = sum(var(delta(ptr:ptr+Flen-1,:),1,2));
    wF = 1/max(varF, eps);
    WF(:,:) = wF * eye(Flen);
    ptr = ptr + Flen;
    nFinfo = 1;
end

ptr_out = ptr;
end

function [WC, nCinfo] = build_WC(delta, ptr_in, offset, layout, Clen)
% Build weights for C block: 1 info per scalar condition (row-by-row)
% Returns WC [Clen x Clen] diagonal.

WC = [];
nCinfo = 0;

if Clen <= 0
    return;
end

WC = zeros(Clen, Clen);

if isfield(layout,'C') && isfield(layout.C,'items') && ~isempty(layout.C.items)
    ptr = ptr_in;
    local = 1;
    for i=1:numel(layout.C.items)
        it = layout.C.items(i);
        for c=1:it.nCond
            rr_delta = ptr; % one row in C
            v = var(delta(rr_delta,:),1,2);
            w = 1/max(v, eps);

            WC(local, local) = w;

            ptr = ptr + 1;
            local = local + 1;
            nCinfo = nCinfo + 1;
        end
    end
else
    % fallback: treat whole C as 1 info
    varC = sum(var(delta(ptr_in:ptr_in+Clen-1,:),1,2));
    wC = 1/max(varC, eps);
    WC(:,:) = wC * eye(Clen);
    nCinfo = 1;
end
end

function stats = pack_stats(mode, Dlen, Ulen, Flen, Clen, nDinfo, nFinfo, nCinfo)
stats = struct();
stats.mode = mode;
stats.sizes = struct('D',Dlen,'U',Ulen,'F',Flen,'C',Clen);
stats.ninfo = struct('D',nDinfo,'F',nFinfo,'C',nCinfo,'total',nDinfo+nFinfo+nCinfo);
end

File: python/src/pme_toolkit/weights.py
from __future__ import annotations

from typing import Any

import numpy as np
from numpy.typing import NDArray


Array = NDArray[np.float64]


def _safe_invv(x: float) -> float:
    return 1.0 / max(float(x), np.finfo(float).eps)


def _sum_row_variances(block: Array) -> float:
    if block.size == 0:
        return 0.0
    return float(np.sum(np.var(block, axis=1, ddof=0)))


def _pack_stats(
    mode: str,
    dlen: int,
    ulen: int,
    flen: int,
    clen: int,
    n_dinfo: int,
    n_finfo: int,
    n_cinfo: int,
) -> dict[str, Any]:
    return {
        "mode": mode,
        "sizes": {"D": dlen, "U": ulen, "F": flen, "C": clen},
        "ninfo": {
            "D": n_dinfo,
            "F": n_finfo,
            "C": n_cinfo,
            "total": n_dinfo + n_finfo + n_cinfo,
        },
    }


def build_wf(
    delta: Array,
    ptr_in: int,
    layout: dict[str, Any],
    flen: int,
) -> tuple[Array, int, int]:
    """
    Build weights for F block: 1 information per field condition.

    Python port of MATLAB build_WF(delta, ptr_in, offset, layout, Flen).

    Parameters
    ----------
    delta
        Delta matrix in P-space, shape [Np x S]
    ptr_in
        0-based starting row in delta for the F block
    layout
        Parsed layout dictionary
    flen
        Number of F rows in P

    Returns
    -------
    wf
        [flen x flen] diagonal block
    n_finfo
        Number of field information sources
    ptr_out
        Updated 0-based row pointer in delta
    """
    if flen <= 0:
        return np.zeros((0, 0), dtype=float), 0, ptr_in

    wf = np.zeros((flen, flen), dtype=float)
    n_finfo = 0
    ptr = int(ptr_in)

    f_layout = dict(layout.get("F", {}) or {})
    items = list(f_layout.get("items", []) or [])

    if items:
        local = 0
        for item in items:
            n_cond = int(item["nCond"])
            n_rows = int(item["nRows"])
            k = int(n_rows // max(n_cond, 1))

            for _ in range(n_cond):
                rr_delta = slice(ptr, ptr + k)
                var_block = _sum_row_variances(delta[rr_delta, :])
                w = _safe_invv(var_block)

                rr_w = slice(local, local + k)
                wf[rr_w, rr_w] = w * np.eye(k, dtype=float)

                ptr += k
                local += k
                n_finfo += 1
    else:
        # fallback: treat whole F block as 1 information source
        rr_delta = slice(ptr, ptr + flen)
        var_f = _sum_row_variances(delta[rr_delta, :])
        w_f = _safe_invv(var_f)
        wf[:, :] = w_f * np.eye(flen, dtype=float)
        ptr += flen
        n_finfo = 1

    return wf, n_finfo, ptr


def build_wc(
    delta: Array,
    ptr_in: int,
    layout: dict[str, Any],
    clen: int,
) -> tuple[Array, int]:
    """
    Build weights for C block: 1 information per scalar condition.

    Python port of MATLAB build_WC(delta, ptr_in, offset, layout, Clen).

    Parameters
    ----------
    delta
        Delta matrix in P-space, shape [Np x S]
    ptr_in
        0-based starting row in delta for the C block
    layout
        Parsed layout dictionary
    clen
        Number of C rows in P

    Returns
    -------
    wc
        [clen x clen] diagonal block
    n_cinfo
        Number of scalar information sources
    """
    if clen <= 0:
        return np.zeros((0, 0), dtype=float), 0

    wc = np.zeros((clen, clen), dtype=float)
    n_cinfo = 0

    c_layout = dict(layout.get("C", {}) or {})
    items = list(c_layout.get("items", []) or [])

    if items:
        ptr = int(ptr_in)
        local = 0
        for item in items:
            n_cond = int(item["nCond"])
            for _ in range(n_cond):
                rr_delta = ptr
                v = float(np.var(delta[rr_delta, :], ddof=0))
                w = _safe_invv(v)

                wc[local, local] = w

                ptr += 1
                local += 1
                n_cinfo += 1
    else:
        # fallback: treat whole C block as 1 information source
        rr_delta = slice(ptr_in, ptr_in + clen)
        var_c = _sum_row_variances(delta[rr_delta, :])
        w_c = _safe_invv(var_c)
        wc[:, :] = w_c * np.eye(clen, dtype=float)
        n_cinfo = 1

    return wc, n_cinfo


def build_weights(
    delta: Array,
    layout: dict[str, Any],
    cfg: dict[str, Any],
    uinfo: dict[str, Any],
    blocks: dict[str, Array],
) -> tuple[Array, dict[str, Any]]:
    """
    Build block-diagonal weights W consistent with 'ninf as #information'.

    Faithful Python port of MATLAB +pme/weights.m.

    Idea:
    - each INFORMATION contributes ~1 to trace(W*Cov)
    - geometry D counts as 1 info (when present in P)
    - each field condition counts as 1 info
    - each scalar condition counts as 1 info
    - variables U weight is usually 0

    Parameters
    ----------
    delta
        P - P0, shape [Np x S]
    layout
        Parsed layout
    cfg
        Case configuration
    uinfo
        Variable info dictionary
    blocks
        Dict with raw blocks {"D","Ubase","F","C"} already sliced from DB_used

    Returns
    -------
    W
        Weight matrix [Np x Np]
    stats
        Weight diagnostics consistent with MATLAB
    """
    mode = str(cfg["mode"]).lower()

    dlen = int(layout["D"]["nRows"])
    ulen = int(uinfo["Mact"])
    flen = int(blocks["F"].shape[0]) if "F" in blocks else 0
    clen = int(blocks["C"].shape[0]) if "C" in blocks else 0

    if mode == "pme":
        # P = [D; U]
        n = dlen + ulen
        w = np.zeros((n, n), dtype=float)

        var_d = _sum_row_variances(delta[0:dlen, :])
        w_d = _safe_invv(var_d)
        w[0:dlen, 0:dlen] = w_d * np.eye(dlen, dtype=float)

        # U usually not weighted in PME
        w_u = 0.0
        if w_u > 0.0:
            a = dlen
            b = dlen + ulen
            w[a:b, a:b] = w_u * np.eye(ulen, dtype=float)

        stats = _pack_stats(mode, dlen, ulen, flen, clen, 1, 0, 0)
        stats["wD"] = w_d
        stats["wU"] = w_u
        stats["varD"] = var_d
        stats["varU"] = _sum_row_variances(delta[dlen:dlen + ulen, :])

        return w, stats

    if mode == "pi":
        # P = [D; U; F; C]
        n = dlen + ulen + flen + clen
        w = np.zeros((n, n), dtype=float)

        ptr = 0

        # --- D = 1 information ---
        var_d = _sum_row_variances(delta[ptr:ptr + dlen, :])
        w_d = _safe_invv(var_d)
        w[0:dlen, 0:dlen] = w_d * np.eye(dlen, dtype=float)
        ptr += dlen

        # --- U typically 0 ---
        var_u = _sum_row_variances(delta[ptr:ptr + ulen, :])
        ptr_u0 = ptr
        ptr += ulen

        w_u = 0.0
        if w_u > 0.0:
            a = dlen
            b = dlen + ulen
            w[a:b, a:b] = w_u * np.eye(ulen, dtype=float)

        # --- F: 1 info per field condition ---
        wf, n_finfo, ptr = build_wf(delta, ptr, layout, flen)
        if wf.size > 0:
            a = dlen + ulen
            b = dlen + ulen + flen
            w[a:b, a:b] = wf

        # --- C: 1 info per scalar condition ---
        wc, n_cinfo = build_wc(delta, ptr, layout, clen)
        if wc.size > 0:
            a = dlen + ulen + flen
            b = dlen + ulen + flen + clen
            w[a:b, a:b] = wc

        stats = _pack_stats(mode, dlen, ulen, flen, clen, 1, n_finfo, n_cinfo)
        stats["ptrU0"] = ptr_u0
        stats["wD"] = w_d
        stats["wU"] = w_u
        stats["varD"] = var_d
        stats["varU"] = var_u

        return w, stats

    if mode == "pd":
        # P = [U; F; C]
        n = ulen + flen + clen
        w = np.zeros((n, n), dtype=float)

        # --- U weight 0 ---
        w_u = 0.0
        if w_u > 0.0:
            w[0:ulen, 0:ulen] = w_u * np.eye(ulen, dtype=float)

        var_u = _sum_row_variances(delta[0:ulen, :])

        ptr = ulen

        # --- F: 1 info per field condition ---
        wf, n_finfo, ptr = build_wf(delta, ptr, layout, flen)
        if wf.size > 0:
            a = ulen
            b = ulen + flen
            w[a:b, a:b] = wf

        # --- C: 1 info per scalar condition ---
        wc, n_cinfo = build_wc(delta, ptr, layout, clen)
        if wc.size > 0:
            a = ulen + flen
            b = ulen + flen + clen
            w[a:b, a:b] = wc

        stats = _pack_stats(mode, dlen, ulen, flen, clen, 0, n_finfo, n_cinfo)
        stats["wU"] = w_u
        stats["varU"] = var_u

        if flen == 0:
            stats["warning"] = "PD-PME with no F: physics-driven DR limited by C."

        return w, stats

    raise ValueError(f"Unknown mode: {mode}")
